ENH Add Array API compatibility to MinMaxScaler

betatim opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/main/CONTRIBUTING.md
-->

#### Reference Issues/PRs
<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->
Towards #26024


#### What does this implement/fix? Explain your changes.
This enables `MinMaxScaler` to work with Array API compatible arrays. Most of the changes are replacing `np` with `xp` (which represents the namespace the array the user passed in belongs to). Had to implement some helpers like `nanmin` and `nanmax`.

#### Any other comments?


<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

betatim committed:
Add Array API compatibility to MinMaxScaler

thomasjpfan commented on sklearn/utils/_array_api.py:
```diff
@@ -357,6 +357,28 @@ def _expit(X):
+def _nanmin(X, axis=None):
+    xp, _ = get_namespace(X)
+    if _is_numpy_namespace(xp):
+        return numpy.nanmin(X, axis=axis)
```
Unfortunately, this still needs `xp.asarray`

```suggestion
        return xp.asarray(numpy.nanmin(X, axis=axis))
```

At this point, `xp` can be a `numpy.array_api` Array which means the output needs to be one too.

thomasjpfan reviewed (commented):
This looks fairly straight forward. We'll need tests to make sure PyTorch works on the CI and gives correct results.

For completeness, `inverse_transform` can also be updated with `array_api`.

betatim committed:
Ensure returned arrays are in the same namespace

betatim committed:
Add what's new

betatim commented:
Done and done.

For the tests I more or less took the tests that exist for `LinearDiscriminantAnalysis`. They are quite similar but not identical. I was thinking I'll make a new PR that tries to make a set of "common tests". Probably via adding a new estimator tag that can be used to discover things that support the array API. WDYT?

betatim committed:
Fix lint

betatim committed:
Fix author attribution

betatim commented on sklearn/preprocessing/_data.py:
```diff
@@ -471,22 +475,24 @@ def partial_fit(self, X, y=None):
             X,
             reset=first_pass,
-            dtype=FLOAT_DTYPES,
+            dtype=(xp.float64, xp.float32, xp.float16),
```
What shall we do about the fact that `float16` is not part of the Array API? I assume it is important to have it here (even if we did remove it, it would mean a deprecation cycle). Should we add it to a scikit-learn internal "compat layer"? `float16` seems to exist in Numpy, Torch and cupy.

Thoughts?

betatim replied:
For torch (not checked cupy) `xp.float16` works/exists. I think as part of the Array API standard it is allowed to add additional dtypes (in this case float16), but `numpy.array_api` only contains the required types.

betatim committed:
Make capitalization of Array API consistent

betatim committed:
Add float16 to Numpy Array API namespace

The float16 dtype is not required by the spec, but the spec allows us to
add additional dtypes. Other Array API implementations (torch) support
the flaot16 dtype already.

betatim committed:
Add comment about namespace wrapping

thomasjpfan commented on sklearn/utils/_array_api.py:
```diff
@@ -139,6 +139,13 @@ def __init__(self, array_namespace):
         return getattr(self._namespace, name)
 
+    @property
+    def float16(self):
```
If we want to support `float16`, then we also need to update `isdtype`:

```patch
diff --git a/sklearn/utils/_array_api.py b/sklearn/utils/_array_api.py
index 89560b9370..3371fa6c92 100644
--- a/sklearn/utils/_array_api.py
+++ b/sklearn/utils/_array_api.py
@@ -98,7 +98,10 @@ def _isdtype_single(dtype, kind, *, xp):
                 for k in ("signed integer", "unsigned integer")
             )
         elif kind == "real floating":
-            return dtype in {xp.float32, xp.float64}
+            floats_dtypes = {xp.float32, xp.float64}
+            if hasattr(xp, "float16"):
+                floats_dtypes.add(xp.float16)
+            return dtype in floats_dtypes
         elif kind == "complex floating":
             # Some name spaces do not have complex, such as cupy.array_api
             # and numpy.array_api
@@ -142,7 +145,7 @@ class _ArrayAPIWrapper:
     @property
     def float16(self):
         if self._namespace.__name__ == "numpy.array_api":
-            return numpy.float16
+            return numpy.dtype(numpy.float16)
         else:
             return self._namespace.float16
 
@@ -171,7 +174,7 @@ class _ArrayAPIWrapper:
         return self._namespace.stack(selected, axis=axis)
 
     def isdtype(self, dtype, kind):
-        return isdtype(dtype, kind, xp=self._namespace)
+        return isdtype(dtype, kind, xp=self)
 
 
 def _check_device_cpu(device):  # noqa
@@ -225,6 +228,7 @@ class _NumPyAPIWrapper:
         "uint16",
         "uint32",
         "uint64",
+        "float16",
         "float32",
         "float64",
         "complex64",
diff --git a/sklearn/utils/tests/test_array_api.py b/sklearn/utils/tests/test_array_api.py
index 77fa20e6d0..b941d207d4 100644
--- a/sklearn/utils/tests/test_array_api.py
+++ b/sklearn/utils/tests/test_array_api.py
@@ -253,6 +253,7 @@ def test_get_namespace_array_api_isdtype(wrapper):
     assert xp.isdtype(xp.float32, xp.float32)
     assert xp.isdtype(xp.float32, "real floating")
     assert xp.isdtype(xp.float64, "real floating")
+    assert xp.isdtype(xp.float16, "real floating")
     assert not xp.isdtype(xp.int32, "real floating")
 
     assert xp.isdtype(xp.bool, "bool")

```

betatim replied:
I added a variation of this, based on the `supported_float_types()` below. Because some Array API implementations support the additional `flaot16` type

thomasjpfan commented on sklearn/preprocessing/_data.py:
```diff
@@ -514,18 +520,20 @@ def transform(self, X):
             X,
             copy=self.copy,
-            dtype=FLOAT_DTYPES,
+            dtype=(xp.float64, xp.float32, xp.float16),
```
If there is a another Array API implementation that does not have `float16`, I do not want our Array API wrapper to break. The Array API spec only enforces that `float32` and `float64` exists.

```python
def _supported_float_dtypes(xp):
    if hasattr(xp, "float16"):
        return {xp.float16, xp.float32, xp.float64}
    else:
        return {xp.float32, xp.float64}
```

betatim replied:
This is a smart idea. Done.

thomasjpfan reviewed (commented):
I am uneasy about supporting `xp.float16`. Adding a `float16` property to `_ArrayAPIWrapper` feels like we are "extending the Array API spec".

My preference is to be strict and only allow `float16` when the arrays are `ndarray`. This means we do not support `float16` for `numpy.array_api` Arrays, `torch`, `cupy.array_api`, etc.

betatim committed:
Don't add float16 to Numpy Array API namespace

betatim commented:
> My preference is to be strict and only allow `float16` when the arrays are `ndarray`. This means we do not support `float16` for `numpy.array_api` Arrays, `torch`, `cupy.array_api`, etc.

Fine for me. Though as far as I can see `numpy.array_api` and `cupy.array_api` will be the only namespaces for which we won't have `float16`. For `cupy`, `torch` and `numpy` it exists.

thomasjpfan commented on sklearn/utils/_array_api.py:
```diff
@@ -98,7 +98,7 @@ def _isdtype_single(dtype, kind, *, xp):
             )
         elif kind == "real floating":
-            return dtype in {xp.float32, xp.float64}
+            return dtype in supported_float_dtypes(xp)
```
Can we add a test for this and update add the type in`_NumPyAPIWrapper`:

```patch
diff --git a/sklearn/utils/_array_api.py b/sklearn/utils/_array_api.py
index 852713ce7f..6a7bacc6b7 100644
--- a/sklearn/utils/_array_api.py
+++ b/sklearn/utils/_array_api.py
@@ -226,6 +226,7 @@ class _NumPyAPIWrapper:
         "uint16",
         "uint32",
         "uint64",
+        "float16",
         "float32",
         "float64",
         "complex64",
diff --git a/sklearn/utils/tests/test_array_api.py b/sklearn/utils/tests/test_array_api.py
index 77fa20e6d0..841a274425 100644
--- a/sklearn/utils/tests/test_array_api.py
+++ b/sklearn/utils/tests/test_array_api.py
@@ -255,6 +255,9 @@ def test_get_namespace_array_api_isdtype(wrapper):
     assert xp.isdtype(xp.float64, "real floating")
     assert not xp.isdtype(xp.int32, "real floating")
 
+    if isinstance(xp, _NumPyAPIWrapper):
+        assert xp.isdtype(xp.float16, "real floating")
+
     assert xp.isdtype(xp.bool, "bool")
     assert not xp.isdtype(xp.float32, "bool")
```

betatim replied:
I did something slightly different for the test. Used `supported_float_dtypes` instead of testing `float16` only for Numpy.

thomasjpfan commented on sklearn/utils/_array_api.py:
```diff
@@ -119,14 +119,22 @@ def _isdtype_single(dtype, kind, *, xp):
+def supported_float_dtypes(xp):
+    """Supported floating point types for the namespace"""
+    if hasattr(xp, "float16"):
+        return (xp.float64, xp.float32, xp.float16)
+    else:
+        return (xp.float64, xp.float32)
```
Nit: 

```suggestion
        return {xp.float64, xp.float32, xp.float16}
    else:
        return {xp.float64, xp.float32}
```

ogrisel replied:
I am pretty sure that tuples are always faster / less overhead for most operations when they have low cardinality. But fair enough, I am fine with either :)

thomasjpfan replied:
Okay, lets go with tuples.

thomasjpfan commented on sklearn/utils/_array_api.py:
```diff
@@ -357,6 +367,28 @@ def _expit(X):
+    else:
+        X = xp.max(xp.where(~xp.isnan(X), X, xp.asarray(-xp.inf)), axis=axis)
+        # Replace Infs from all NaN slices with NaN again
+        return xp.where(~xp.isneginf(X), X, xp.asarray(xp.nan))
```
`isneginf` is not a part of the Array API spec and will fail with `cupy.array_api`. `xp.isneginf` will also fail for Array libraries that strictly follow the spec.

thomasjpfan reviewed (commented):
Okay, I am mostly onboard with supporting `float16`.

ogrisel commented on sklearn/preprocessing/tests/test_data.py:
```diff
@@ -2654,3 +2657,113 @@ def test_kernel_centerer_feature_names_out():
+@skip_if_array_api_compat_not_configured
+@pytest.mark.parametrize("device", ["cuda", "cpu"])
+@pytest.mark.parametrize("dtype", ["float16", "float32", "float64"])
+def test_minmax_array_torch(device, dtype):
```
I think we need a generic way to parametrize such tests with a list of array api libraries that we want to officially support. For now I would always test with cupy and pytorch by default (skipped when not installed).

We need some pytest helpers factorize the boiler plate, in particular because some namespace containers support multiple device options and other don't.

ogrisel reviewed (commented):
Besides the following and @thomasjpfan's comment on `isneginf`, LGTM.

Not sure how to efficiently cope with the lack of a generic `isneginf` though.

betatim committed:
Use our own isneginf function

betatim committed:
Add float16 to Numpy wrapper, add float dtype test

ogrisel commented on sklearn/utils/_array_api.py:
```diff
@@ -357,6 +368,35 @@ def _expit(X):
+    return xp.logical_and(is_inf, signbit)
+
+
+def _nanmin(X, axis=None):
```
```suggestion
def _nanmin(X, axis=None):
    # TODO: refactor once nan-aware reductions are standardized:
    # https://github.com/data-apis/array-api/issues/621
```

ogrisel reviewed (approved):
Beyond making `atol` stricter in the tests and the following suggestions to improve comments / docstrings, LGTM!

betatim commented on sklearn/utils/_array_api.py:
```diff
@@ -357,6 +368,35 @@ def _expit(X):
     return 1.0 / (1.0 + xp.exp(-X))
 
 
+def _isneginf(X):
```
TODO for Tim: open issue in Array API about asymmetry between `isinf` and `isneginf`

ogrisel commented:
I think this PR needs to be updated to leverage the new common test recently merged in `main`.

betatim committed:
Ruffing it up

betatim committed:
Fix author

betatim committed:
Add tests for NaN reductions

betatim committed:
Add comments about special status of float16

betatim committed:
Change library order

betatim committed:
Enable array API when testing

betatim commented:
I think this is ready to go now. I'd like to add more tests a la:
```python
@parametrize_with_checks(
    [
        MinMaxScaler()
    ],
    check_yielder=_yield_array_api_checks,
)
def test_array_api_compliance(check, estimator, request):
    check(estimator, check_values=True)
```
but this uses things from #26315. Either we wait to merge this PR until #26315 has landed or we make another PR later to add those tests. I'm fine with either.

ogrisel commented:
I just reverted the change made to `parametrize_with_checks` because in the end I no longer needed it #26315. Instead I used a cross-product of regular `pytest.mark.parametrize`:

https://github.com/scikit-learn/scikit-learn/pull/26315/files#diff-7fdbfa7b6564912adf698653f0faa9248c7ad600cdccf607451b5c500dbf6b4cR731-R755

Note: the order of the decorator matters for the name of the tests.

betatim committed:
Formatting

ogrisel commented on sklearn/preprocessing/tests/test_data.py:
```diff
@@ -684,6 +691,26 @@ def test_standard_check_array_of_inverse_transform():
 
 
+@pytest.mark.parametrize(
+    "array_namepsace, device, dtype", yield_namespace_device_dtype_combinations()
```
typo: namespace

ogrisel reviewed (approved):
Thanks for the update. Still LGTM.

betatim committed:
Fix typo

betatim commented:
@thomasjpfan what do you think?

thomasjpfan commented on sklearn/utils/_array_api.py:
```diff
@@ -460,6 +480,39 @@ def _weighted_sum(sample_score, sample_weight, normalize=False, xp=None):
+    else:
+        X = xp.min(xp.where(~xp.isnan(X), X, xp.asarray(+xp.inf)), axis=axis)
+        # Replace Infs from all NaN slices with NaN again
+        return xp.where(~xp.isinf(X), X, xp.asarray(xp.nan))
```
This would replace a valid `inf` if the input was actually all inf:

```python
import numpy as np

X = np.array([np.inf, np.inf])
np.nanmin(X)
# inf
```

betatim replied:
Fixed. The code now is much better I think, and it has tests!

thomasjpfan commented on sklearn/utils/_array_api.py:
```diff
@@ -460,6 +480,39 @@ def _weighted_sum(sample_score, sample_weight, normalize=False, xp=None):
+        return xp.asarray(numpy.nanmin(X, axis=axis))
+
+    else:
+        X = xp.min(xp.where(~xp.isnan(X), X, xp.asarray(+xp.inf)), axis=axis)
```
Instead of doing the negation and creating another array, we can flip the condition:

```suggestion
        X = xp.min(xp.where(xp.isnan(X), xp.asarray(+xp.inf), X), axis=axis)
```

thomasjpfan commented on sklearn/utils/_array_api.py:
```diff
@@ -460,6 +480,39 @@ def _weighted_sum(sample_score, sample_weight, normalize=False, xp=None):
+        return xp.asarray(numpy.nanmax(X, axis=axis))
+
+    else:
+        X = xp.max(xp.where(~xp.isnan(X), X, xp.asarray(-xp.inf)), axis=axis)
```
Same here regarding flipping the negation:

```suggestion
        X = xp.max(xp.where(xp.isnan(X), xp.asarray(-xp.inf), X), axis=axis)
```

thomasjpfan reviewed (commented):
Can you add the estimator to:

https://github.com/scikit-learn/scikit-learn/blob/2b0eef892d4ee973789089ac82f24669a89a3859/doc/modules/array_api.rst?plain=1#L88-L89

I ran this on my local machine with CuPy + PyTorch and the new tests pass.

betatim committed:
Move changelog entry

betatim committed:
Fix all NaN slice handling

When using the axis keyword it can happen that a slice contains only
NaNs. This change corrects the logic that restores the NaNs at the end.
Added tests for 2D inputs.

betatim committed:
Remove isneginf() as it is unused

thomasjpfan commented on sklearn/utils/_array_api.py:
```diff
@@ -460,6 +480,40 @@ def _weighted_sum(sample_score, sample_weight, normalize=False, xp=None):
+        X = xp.min(xp.where(mask, xp.asarray(+xp.inf), X), axis=axis)
+        # Replace Infs from all NaN slices with NaN again
+        mask = xp.all(mask, axis=axis)
+        if xp.any(mask):
+            X = xp.where(~mask, X, xp.asarray(xp.nan))
```
Inverting the `where` clause means we no longer need to create a second `~mask`:

```suggestion
        if xp.any(mask):
            X = xp.where(mask, xp.asarray(xp.nan), X)
```

thomasjpfan commented on sklearn/utils/_array_api.py:
```diff
@@ -460,6 +480,40 @@ def _weighted_sum(sample_score, sample_weight, normalize=False, xp=None):
+        # Replace Infs from all NaN slices with NaN again
+        mask = xp.all(mask, axis=axis)
+        if xp.any(mask):
+            X = xp.where(~mask, X, xp.asarray(xp.nan))
```
Same here:

```suggestion
            X = xp.where(mask, xp.asarray(xp.nan), X)
```

thomasjpfan reviewed (approved):
Minor comment, otherwise LGTM

betatim committed:
Change order to avoid having to create a duplicate mask

betatim committed:
Merge remote-tracking branch 'upstream/main' into minmax-array-api

thomasjpfan committed:
Merge remote-tracking branch 'upstream/main' into pr/26243

betatim committed:
Empty-Commit

betatim committed:
Merge remote-tracking branch 'origin/minmax-array-api' into minmax-array-api
